app/controllers/application_controller.rb

```diff
@@ -5,6 +5,176 @@ class ApplicationController < Sinatra::Base
+        session[:user_id] = @user.id
+      
+        redirect '/tweets'
+      else 
+        redirect '/signup'
+      
+    end 
+  end 
+
+  get '/tweets' do 
+ 
+    if is_logged_in?
+        @fweets = Tweet.all
+        @posts = []
+        @fweets.each do |f|
+        @posts << {username: User.find(f.user_id).username, content: f.content, user_id: f.user_id, id: f.id}
+          end
+      @user = User.find(session[:user_id])
+      erb :'tweets/tweets' 
+    else
+     redirect '/login'
+    end 
+  end 
+
+  get '/tweets/new' do 
+      if is_logged_in?
+        erb :'/tweets/create_tweet'
+      else 
+        redirect "/login"
+    end
+  end 
+
+  post '/tweets/new' do 
+   if !params[:tweet].present?
+    redirect '/tweets/new'
+  end 
+
+    if is_logged_in? && params[:tweet].present?
+      @user = User.find(session[:user_id])
+      @tweet = Tweet.create(content: params[:tweet], user_id: @user.id)
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end 
+  end 
+
+
+  post '/login' do
+    @user = User.find_by(:username => params[:username])
+        if @user && @user.authenticate(params[:password])
+              session[:user_id] = @user.id
+        redirect '/tweets'
+      else 
+      erb :'/users/login'
+    end 
+  end 
+
+  get '/tweets/:id' do 
+   
+      if is_logged_in?
+        @tweet = Tweet.find(params[:id])
+     
+        erb :'/tweets/one_tweet'
+      else 
+        redirect '/login'
+      end
+    end
+
+  get '/logout' do 
+    session.clear 
+    redirect '/login'
+  end 
+
+  get '/users/:slug' do 
+    
+    @user = User.find_by(username: params[:slug])
+      @fweets = Tweet.where(user_id: @user.id)
+      erb :'/tweets/show_tweet'
+  end 
+
+
+  get '/tweets/:id/edit' do 
+    binding.pry
+    if is_logged_in? && User.find(Tweet.find(params[:id]).user_id) == User.find(session[:user_id])
+    @tweet = Tweet.find(params[:id])
+    erb :'tweets/edit_tweet'
+  else 
+      redirect '/login'
+    end 
+  end 
+
+  post '/tweets/:id/edit' do 
+  
+    
+  if params[:tweet].present?
+    @tweet= Tweet.find(params[:id])
+    @tweet.update(content: params[:tweet])
+
+    redirect '/tweets'
+  else 
+    redirect "/tweets/#{@tweet.id}/edit"
+  end 
+  end
+
+
+  get '/tweet/:id/delete' do 
+    
+  if is_logged_in? && User.find(Tweet.find(params[:id]).user_id) == User.find(session[:user_id])
+    Tweet.find(params[:id]).delete
+    redirect '/tweets'
+  else 
+    redirect '/tweets'
+  end 
+  end 
+
+  post '/tweet/:id/delete' do 
+  
+    if is_logged_in? && User.find(Tweet.find(params[:id]).user_id) == User.find(session[:user_id])
+      @tweet = Tweet.find(params[:id])
+
+    redirect "/tweet/#{@tweet.id}/delete"
+  else 
+    redirect '/tweets'
+  end 
+  end 
+
+
+  
+  helpers do 
+
+    def current_user
+      User.find(session["user_id"])
+    end 
+
+    def is_logged_in?
+       !!session[:user_id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```